<!DOCTYPE html>
<html>
<head><meta name="generator" content="Hexo 3.9.0">
  <meta charset="utf-8">
  <!-- Google Adsense -->
  <script async data-ad-client="ca-pub-8106207149031244" src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script>
  <!-- End Google Adsense -->
  
<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-08PF7150CV"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'G-08PF7150CV');
</script>
<!-- End Google Analytics -->

  
<!-- Google Tag Manager -->
<script>(function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
  new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
  j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
  'https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);
  })(window,document,'script','dataLayer','GTM-N29SVMB');</script>
<!-- End Google Tag Manager -->

  
  <title>C++ exception 例外處理 | ShengYu Talk</title>
  <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
  <meta name="description" content="本篇介紹 c++ exception 例外處理，程式在執行過程中可能會發生一些特殊情況的異常，例如：記憶體不夠、越界存取，所以我們可以用 try catch 來處理這種例外情形   C++ 基本的 try… catch… 例外處理最簡單的例外處理的寫法如下，使用 try... catch... 所組成，用 try 包住可能會發生例外的程式碼區段，用 catch 來處理這個 try 區域所發生的例外">
<meta name="keywords" content="C&#x2F;C++">
<meta property="og:type" content="article">
<meta property="og:title" content="C++ exception 例外處理">
<meta property="og:url" content="https://shengyu7697.github.io/cpp-exception/index.html">
<meta property="og:site_name" content="ShengYu Talk">
<meta property="og:description" content="本篇介紹 c++ exception 例外處理，程式在執行過程中可能會發生一些特殊情況的異常，例如：記憶體不夠、越界存取，所以我們可以用 try catch 來處理這種例外情形   C++ 基本的 try… catch… 例外處理最簡單的例外處理的寫法如下，使用 try... catch... 所組成，用 try 包住可能會發生例外的程式碼區段，用 catch 來處理這個 try 區域所發生的例外">
<meta property="og:locale" content="zh-TW">
<meta property="og:updated_time" content="2024-05-18T01:36:17.000Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="C++ exception 例外處理">
<meta name="twitter:description" content="本篇介紹 c++ exception 例外處理，程式在執行過程中可能會發生一些特殊情況的異常，例如：記憶體不夠、越界存取，所以我們可以用 try catch 來處理這種例外情形   C++ 基本的 try… catch… 例外處理最簡單的例外處理的寫法如下，使用 try... catch... 所組成，用 try 包住可能會發生例外的程式碼區段，用 catch 來處理這個 try 區域所發生的例外">
  
    <link rel="alternate" href="/atom.xml" title="ShengYu Talk" type="application/atom+xml">
  
  
    <link rel="icon" href="/favicon.png">
  
  
    <link href="//fonts.googleapis.com/css?family=Source+Code+Pro" rel="stylesheet" type="text/css">
  
  <link rel="stylesheet" href="/css/style.css">
</head>

<body>

<!-- Google Tag Manager (noscript) -->
<noscript><iframe src="https://www.googletagmanager.com/ns.html?id=GTM-N29SVMB"
height="0" width="0" style="display:none;visibility:hidden"></iframe></noscript>
<!-- End Google Tag Manager (noscript) -->

  <div id="container">
    <div id="wrap">
      <header id="header">
  <div id="banner"></div>
  <div id="header-outer" class="outer">
    <div id="header-title" class="inner">
      <h1 id="logo-wrap">
        <a href="/" id="logo">ShengYu Talk</a>
      </h1>
      
    </div>
    <div id="header-inner" class="inner">
      <nav id="main-nav">
        <a id="main-nav-toggle" class="nav-icon"></a>
        
          <a class="main-nav-link" href="/">首頁</a>
        
          <a class="main-nav-link" href="/categories/Python教學/">Python教學</a>
        
          <a class="main-nav-link" href="/categories/C-C-教學/">C/C++教學</a>
        
          <a class="main-nav-link" href="/archives">歸檔</a>
        
          <a class="main-nav-link" href="/tags">標籤</a>
        
          <a class="main-nav-link" href="/projects">Projects</a>
        
          <a class="main-nav-link" href="/about">關於我</a>
        
      </nav>
      <nav id="sub-nav">
        <a id="nav-linkedin-link" class="nav-icon" href="https://www.linkedin.com/in/shengyu-peng-48b664a6" rel="external nofollow noreferrer" title="Linkedin"></a>
        <a id="nav-facebook-link" class="nav-icon" href="https://www.facebook.com/shengyu7697" rel="external nofollow noreferrer" title="Facebook"></a>
        <a id="nav-github-link" class="nav-icon" href="https://github.com/shengyu7697" rel="external nofollow noreferrer" title="GitHub"></a>
        
          <a id="nav-rss-link" class="nav-icon" href="/atom.xml" title="RSS Feed"></a>
        
        <a id="nav-search-btn" class="nav-icon" title="搜尋"></a>
      </nav>
      <div id="search-form-wrap">
        <form action="//google.com/search" method="get" accept-charset="UTF-8" class="search-form"><input type="search" name="q" class="search-form-input" placeholder="搜尋"><button type="submit" class="search-form-submit">&#xF002;</button><input type="hidden" name="sitesearch" value="https://shengyu7697.github.io"></form>
      </div>
    </div>
  </div>
</header>

      <div class="outer">
        <section id="main"><article id="post-cpp-exception" class="article article-type-post" itemscope itemprop="blogPost">
  <div class="article-meta">
    <a href="/cpp-exception/" class="article-date">
  <time datetime="2020-12-10T14:00:00.000Z" itemprop="datePublished">2020-12-10</time>
</a>
    
  <div class="article-category">
    <a class="article-category-link" href="/categories/C-C-教學/">C/C++教學</a>
  </div>

  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 class="article-title" itemprop="name">
      C++ exception 例外處理
    </h1>
  

      </header>
    
    <div class="article-entry" itemprop="articleBody">
      
        
        <!-- ads-1 -->
        <div id="ads-1">
  <ins class="adsbygoogle"
    style="display:block"
    data-ad-client="ca-pub-8106207149031244"
    data-ad-slot="4194519399"
    data-ad-format="auto"
    data-full-width-responsive="true"></ins>
  <script>
    (adsbygoogle = window.adsbygoogle || []).push({});
  </script>
</div>

        <!-- ads-1 -->
        <p>本篇介紹 c++ exception 例外處理，程式在執行過程中可能會發生一些特殊情況的異常，例如：記憶體不夠、越界存取，所以我們可以用 try catch 來處理這種例外情形  </p>
<h2><span id="c-基本的-try-catch-例外處理">C++ 基本的 try… catch… 例外處理</span></h2><p>最簡單的例外處理的寫法如下，使用 <code>try... catch...</code> 所組成，<br>用 try 包住可能會發生例外的程式碼區段，<br>用 catch 來處理這個 try 區域所發生的例外，<br><code>what()</code> 是 exception 類別提供的一個 public method，exception 類別會被所有子類重載，所以不管發生哪種例外，只要呼叫 exception 類的 <code>what()</code> 方法，就會取得該種例外的發生的原因，這個機制是用 <a href="/cpp-virtual/" title="C++ virtual">C++ virtual</a> 所完成的，程式碼中的 <code>e.what()</code> 將回傳例外產生的原因字串。  </p>
<p>另外，catch 裡面的 exception 變數應該要用 reference 的方式。<br><figure class="highlight cpp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">#<span class="meta-keyword">include</span> <span class="meta-string">&lt;iostream&gt;</span></span></span><br><span class="line"><span class="meta">#<span class="meta-keyword">include</span> <span class="meta-string">&lt;exception&gt;</span></span></span><br><span class="line"></span><br><span class="line"><span class="keyword">using</span> <span class="keyword">namespace</span> <span class="built_in">std</span>;</span><br><span class="line"></span><br><span class="line"><span class="function"><span class="keyword">int</span> <span class="title">main</span><span class="params">()</span></span></span><br><span class="line"><span class="function"></span>&#123;</span><br><span class="line">    <span class="keyword">try</span> &#123;</span><br><span class="line">        <span class="comment">// ...</span></span><br><span class="line">    &#125; <span class="keyword">catch</span> (<span class="built_in">std</span>::exception &amp;e) &#123; <span class="comment">// exception should be caught by reference</span></span><br><span class="line">        <span class="built_in">cout</span> &lt;&lt; <span class="string">"exception: "</span> &lt;&lt; e.what() &lt;&lt; <span class="string">"\n"</span>;</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    <span class="keyword">return</span> <span class="number">0</span>;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure></p>
<p>接下來我們來看看實戰中的例外範例，  </p>
<h2><span id="c-vector-的-out_of_range-例外處理">C++ vector 的 out_of_range 例外處理</span></h2><p>C++ vector 存取元素的方式有兩種，一種是使用 <code>[]</code> operator，例如：<code>v[0] = 0</code>，<br>另一種是使用 <code>at()</code>，例如：<code>v.at(0) = 0</code>，<br>這兩個的差別在於使用 <code>[]</code> operator 存取超界時不會丟出例外 exception，<br>所以像下面這個例子中的 <code>v[3]</code> 實際上是讀取了不知名的記憶體位置的數值，<br>使用 <code>at()</code> 存取超界時會丟出 <code>std::out_of_range</code> 的例外 exception，<br>所以下面這個範例我們嘗試捕捉看看例外，<br><figure class="highlight cpp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">// g++ cpp-exception.cpp -o a.out -std=c++11</span></span><br><span class="line"><span class="meta">#<span class="meta-keyword">include</span> <span class="meta-string">&lt;iostream&gt;</span></span></span><br><span class="line"><span class="meta">#<span class="meta-keyword">include</span> <span class="meta-string">&lt;vector&gt;</span></span></span><br><span class="line"><span class="meta">#<span class="meta-keyword">include</span> <span class="meta-string">&lt;exception&gt;</span></span></span><br><span class="line"></span><br><span class="line"><span class="keyword">using</span> <span class="keyword">namespace</span> <span class="built_in">std</span>;</span><br><span class="line"></span><br><span class="line"><span class="function"><span class="keyword">int</span> <span class="title">main</span><span class="params">()</span></span></span><br><span class="line"><span class="function"></span>&#123;</span><br><span class="line">    <span class="built_in">vector</span>&lt;<span class="keyword">int</span>&gt; v = &#123;<span class="number">1</span>,<span class="number">2</span>,<span class="number">3</span>&#125;;</span><br><span class="line">    <span class="keyword">try</span> &#123;</span><br><span class="line">        <span class="built_in">cout</span> &lt;&lt; v[<span class="number">0</span>] &lt;&lt; <span class="string">"\n"</span>;</span><br><span class="line">        <span class="built_in">cout</span> &lt;&lt; v[<span class="number">1</span>] &lt;&lt; <span class="string">"\n"</span>;</span><br><span class="line">        <span class="built_in">cout</span> &lt;&lt; v[<span class="number">2</span>] &lt;&lt; <span class="string">"\n"</span>;</span><br><span class="line">        <span class="built_in">cout</span> &lt;&lt; v[<span class="number">3</span>] &lt;&lt; <span class="string">"\n"</span>;</span><br><span class="line">        <span class="built_in">cout</span> &lt;&lt; v.at(<span class="number">0</span>) &lt;&lt; <span class="string">"\n"</span>;</span><br><span class="line">        <span class="built_in">cout</span> &lt;&lt; v.at(<span class="number">1</span>) &lt;&lt; <span class="string">"\n"</span>;</span><br><span class="line">        <span class="built_in">cout</span> &lt;&lt; v.at(<span class="number">2</span>) &lt;&lt; <span class="string">"\n"</span>;</span><br><span class="line">        <span class="built_in">cout</span> &lt;&lt; v.at(<span class="number">3</span>) &lt;&lt; <span class="string">"\n"</span>;</span><br><span class="line">    &#125; <span class="keyword">catch</span> (<span class="built_in">std</span>::exception &amp;e) &#123;</span><br><span class="line">        <span class="built_in">cout</span> &lt;&lt; <span class="string">"exception: "</span> &lt;&lt; e.what() &lt;&lt; <span class="string">"\n"</span>;</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    <span class="keyword">return</span> <span class="number">0</span>;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure></p>
<p>程式輸出如下，可以發現 使用 <code>v[3]</code> operator 存取超界用 <code>try... catch...</code> 是捕捉不到例外的，<br>但是當程式執行到 <code>v.at(3)</code> 時存取超界拋出了例外 exception，  </p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">0</span><br><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">exception: vector::_M_range_check: __n (which is 3) &gt;= this-&gt;size() (which is 3)</span><br></pre></td></tr></table></figure>
<h2><span id="常見的-exception-例外">常見的 exception 例外</span></h2><p>以下幾乎是所以有的 exception 例外種類，C++11 以及之後加入的 exception 未列在下面，  </p>
<table>
<thead>
<tr>
<th>例外種類</th>
<th>說明</th>
</tr>
</thead>
<tbody>
<tr>
<td>std::exception</td>
<td>所有標準 C++ 例外的父類</td>
</tr>
<tr>
<td>std::bad_alloc</td>
<td>new 配置記憶體失敗時拋出，可能是記憶體不足</td>
</tr>
<tr>
<td>std::bad_cast</td>
<td>dynamic_cast 失敗時拋出</td>
</tr>
<tr>
<td>std::bad_exception</td>
<td>無法預期的例外</td>
</tr>
<tr>
<td>std::bad_typeid</td>
<td>typeid 拋出</td>
</tr>
<tr>
<td>std::logic_error</td>
<td>程式碼邏輯異常拋出</td>
</tr>
<tr>
<td>std::runtime_error</td>
<td>執行期間的例外</td>
</tr>
<tr>
<td>std::domain_error</td>
<td>Domain error exception</td>
</tr>
<tr>
<td>std::invalid_argument</td>
<td>無效的參數時拋出</td>
</tr>
<tr>
<td>std::length_error</td>
<td>建立長度太長時拋出，例外：std::string</td>
</tr>
<tr>
<td>std::out_of_range</td>
<td>存取越界時拋出，例如：std::vector</td>
</tr>
<tr>
<td>std::overflow_error</td>
<td>溢位時拋出</td>
</tr>
<tr>
<td>std::range_error</td>
<td>Range error exception</td>
</tr>
<tr>
<td>std::underflow_error</td>
<td>算術下溢時拋出</td>
</tr>
</tbody>
</table>
<h2><span id="重新拋出例外-exception">重新拋出例外 exception</span></h2><p>在 C++ 中有時候這個例外 exception 捕捉到了，但是某些原因未去處理或還不想處理，想要將該例外重新拋出，就需要使用 <code>throw</code> 這個關鍵字，<br>重新拋出例外 exception 後，讓該例外交給更外層的例外捕捉處理，<br><figure class="highlight cpp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">try</span> &#123;</span><br><span class="line">    <span class="comment">// ...</span></span><br><span class="line">&#125; <span class="keyword">catch</span> (execption &amp;e) &#123;</span><br><span class="line">    <span class="built_in">cout</span>&lt;&lt; e.what() &lt;&lt; <span class="string">"\n"</span>;</span><br><span class="line">    <span class="keyword">throw</span>; <span class="comment">// 重新拋出例外</span></span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure></p>
<p>參考<br>Exceptions - C++ Tutorials<br><a href="http://www.cplusplus.com/doc/tutorial/exceptions/" rel="noopener external nofollow noreferrer" target="_blank">http://www.cplusplus.com/doc/tutorial/exceptions/</a><br>exception - C++ Reference<br><a href="http://www.cplusplus.com/reference/exception/exception/" rel="noopener external nofollow noreferrer" target="_blank">http://www.cplusplus.com/reference/exception/exception/</a><br>std::exception - cppreference.com<br><a href="https://en.cppreference.com/w/cpp/error/exception" rel="noopener external nofollow noreferrer" target="_blank">https://en.cppreference.com/w/cpp/error/exception</a><br>簡介例外處理<br><a href="https://openhome.cc/Gossip/CppGossip/ExceptionHandling.html" rel="noopener external nofollow noreferrer" target="_blank">https://openhome.cc/Gossip/CppGossip/ExceptionHandling.html</a><br>C++ 异常处理 | 菜鸟教程<br><a href="https://www.runoob.com/cplusplus/cpp-exceptions-handling.html" rel="noopener external nofollow noreferrer" target="_blank">https://www.runoob.com/cplusplus/cpp-exceptions-handling.html</a><br>Throwing out of range exception in C++ - Stack Overflow<br><a href="https://stackoverflow.com/questions/2709719/throwing-out-of-range-exception-in-c/2709733" rel="noopener external nofollow noreferrer" target="_blank">https://stackoverflow.com/questions/2709719/throwing-out-of-range-exception-in-c/2709733</a><br>XYZ的筆記本: C++ 捕捉不到 vector 超出範圍的例外 ?<br><a href="https://xyz.cinc.biz/2013/10/c-vector-try-catch-out-of-range-exception.html" rel="noopener external nofollow noreferrer" target="_blank">https://xyz.cinc.biz/2013/10/c-vector-try-catch-out-of-range-exception.html</a>  </p>
<p><strong>其它相關文章推薦</strong><br><a href="/cpp-tutorial/" title="C/C++ 新手入門教學懶人包">C/C++ 新手入門教學懶人包</a><br><a href="/cpp-virtual/" title="C++ virtual 的兩種用法">C++ virtual 的兩種用法</a><br><a href="/cpp-string-reverse/" title="C/C++ 字串反轉 reverse">C/C++ 字串反轉 reverse</a><br><a href="/cpp-call-by-value-pointer-reference/" title="C/C++ call by value傳值, call by pointer傳址, call by reference傳參考 的差別">C/C++ call by value傳值, call by pointer傳址, call by reference傳參考 的差別</a><br><a href="/cpp-class-template/" title="C++ 類別樣板 class template">C++ 類別樣板 class template</a><br><a href="/std-sort/" title="std::sort 用法與範例">std::sort 用法與範例</a><br><a href="/std-find/" title="std::find 用法與範例">std::find 用法與範例</a><br><a href="/std-queue/" title="std::queue 用法與範例">std::queue 用法與範例</a><br><a href="/std-map/" title="std::map 用法與範例">std::map 用法與範例</a></p>

      
    </div>
    <footer class="article-footer">
      <a data-url="https://shengyu7697.github.io/cpp-exception/" data-id="cm1c3l1wc00b2tmpq5zehp039" data-title="C++ exception 例外處理" class="article-share-link">Share</a>
      
      
  <ul class="article-tag-list"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/C-C/">C/C++</a></li></ul>

    </footer>
  </div>
  
    <!-- ads-2 -->
    <!-- <div id="ads-2">
  <ins class="adsbygoogle"
    style="display:block"
    data-ad-client="ca-pub-8106207149031244"
    data-ad-slot="1792535573"
    data-ad-format="auto"
    data-full-width-responsive="true"></ins>
  <script>
    (adsbygoogle = window.adsbygoogle || []).push({});
  </script>
</div>
 -->
    <!-- ads-2 -->
    <!-- ads-4 -->
    <div id="ads-4">
  <ins class="adsbygoogle"
    style="display:block"
    data-ad-format="autorelaxed"
    data-ad-client="ca-pub-8106207149031244"
    data-ad-slot="5600353687"></ins>
  <script>
    (adsbygoogle = window.adsbygoogle || []).push({});
  </script>
</div>

    <!-- ads-4 -->
    
<nav id="article-nav">
  
    <a href="/linux-proc-memory-status/" id="article-nav-newer" class="article-nav-link-wrap">
      <strong class="article-nav-caption">Newer</strong>
      <div class="article-nav-title">
        
          Linux Android 行程記憶體使用量
        
      </div>
    </a>
  
  
    <a href="/python-tkinter-filedialog-directory/" id="article-nav-older" class="article-nav-link-wrap">
      <strong class="article-nav-caption">Older</strong>
      <div class="article-nav-title">Python tkinter filedialog.askdirectory 選擇資料夾</div>
    </a>
  
</nav>

  
</article>

</section>
        
          <aside id="sidebar">
  
    
  <div class="widget-wrap">
    <h3 class="widget-title">精選文章</h3>
    <div class="widget">
      <ul>
        
          <li><a href="/python-tutorial/">Python 基礎教學目錄</a></li>
        
          <li><a href="/cpp-tutorial/">C/C++ 入門教學目錄</a></li>
        
          <li><a href="/linux-tutorial/">Linux 常用指令教學懶人包</a></li>
        
      </ul>
    </div>
  </div>

  
    
  <div class="widget-wrap">
    <h3 class="widget-title">最新文章</h3>
    <div class="widget">
      <ul>
        
          <li>
            <a href="/dependency-inversion-principle/">設計原則 - 依賴反轉原則 Dependency Inversion Principle</a>
          </li>
        
          <li>
            <a href="/interface-segregation-principle/">設計原則 - 介面隔離原則 Interface Segregation Principle</a>
          </li>
        
          <li>
            <a href="/liskov-substitution-principle/">設計原則 - 里氏替換原則 Liskov Substitution Principle</a>
          </li>
        
          <li>
            <a href="/open-closed-principle/">設計原則 - 開放封閉原則 Open-Closed Principle</a>
          </li>
        
          <li>
            <a href="/single-responsibility-principle/">設計原則 - 單一職責原則 Single Responsibility Principle</a>
          </li>
        
      </ul>
    </div>
  </div>

  
    
  <div class="widget-wrap">
    <h3 class="widget-title">粉絲專頁</h3>
    <iframe src="https://www.facebook.com/plugins/page.php?href=https%3A%2F%2Fwww.facebook.com%2Fshengyu7697&tabs=timeline&width=280&height=500&small_header=true&adapt_container_width=true&hide_cover=false&show_facepile=true&appId" width="280" height="500" style="border:none;overflow:hidden" scrolling="no" frameborder="0" allowfullscreen="true" allow="autoplay; clipboard-write; encrypted-media; picture-in-picture; web-share"></iframe>
  </div>

  
    
  <div class="widget-wrap">
    <h3 class="widget-title">分類</h3>
    <div class="widget">
      <ul class="category-list"><li class="category-list-item"><a class="category-list-link" href="/categories/Arduino/">Arduino</a><span class="category-list-count">4</span></li><li class="category-list-item"><a class="category-list-link" href="/categories/C-C-教學/">C/C++教學</a><span class="category-list-count">221</span></li><li class="category-list-item"><a class="category-list-link" href="/categories/Github/">Github</a><span class="category-list-count">14</span></li><li class="category-list-item"><a class="category-list-link" href="/categories/Git教學/">Git教學</a><span class="category-list-count">20</span></li><li class="category-list-item"><a class="category-list-link" href="/categories/OpenCV教學/">OpenCV教學</a><span class="category-list-count">8</span></li><li class="category-list-item"><a class="category-list-link" href="/categories/PHP教學/">PHP教學</a><span class="category-list-count">14</span></li><li class="category-list-item"><a class="category-list-link" href="/categories/Projects/">Projects</a><span class="category-list-count">6</span></li><li class="category-list-item"><a class="category-list-link" href="/categories/Python教學/">Python教學</a><span class="category-list-count">239</span></li><li class="category-list-item"><a class="category-list-link" href="/categories/Qt教學/">Qt教學</a><span class="category-list-count">7</span></li></ul>
    </div>
  </div>


  
    <div class="widget-wrap" style="position:sticky; position:-webkit-sticky; top:0;">
  <!-- ads-3 -->
  <div id="ads-3">
  <ins class="adsbygoogle"
    style="display:block"
    data-ad-client="ca-pub-8106207149031244"
    data-ad-slot="8896181732"
    data-ad-format="auto"
    data-full-width-responsive="true"></ins>
  <script>
   (adsbygoogle = window.adsbygoogle || []).push({});
  </script>
</div>

  <!-- ads-3 -->
</div>
<script>
  function setSidebarHeight(sidebar) {
    let ad1h = document.getElementById("ads-1").clientHeight; // 280
    //let ad2h = document.getElementById("ads-2").clientHeight; // 280
    let ad3h = document.getElementById("ads-3").clientHeight; // 600
    let ad4h = 0;
    let elem = document.getElementById("ads-4");
    if (elem != null) {
      ad4h = elem.clientHeight; // 447
    }
    let mainh = document.getElementById("main").clientHeight;
    let sidebarh = mainh + ad1h + ad4h - ad3h;
    sidebar.style.height = sidebarh+"px";
    console.log("set sidebar height = " + sidebarh + "px (" + mainh + "+" + ad1h + "+" + ad4h + "-" + ad3h + ")");
  }

  function fixSidebarHeight() {
    let sidebar = document.getElementById("sidebar");
    console.log("sidebar height is " + sidebar.style.height);

    if (sidebar.style.height == "auto") {
      console.log("sidebar height is auto !, fix it.");
      setSidebarHeight(sidebar);
    }

    const observer = new MutationObserver(function (mutations, observer) {
      console.log("observer sidebar height is " + sidebar.style.height);
      if (sidebar.style.height == "auto") {
        setSidebarHeight(sidebar);
      }
    });

    observer.observe(sidebar, {
      attributes: true,
      attributeFilter: ['style']
    });
  }

  console.log("width = " + window.innerWidth);
  if (window.innerWidth >= 768) {
    console.log("width >= 768");
    fixSidebarHeight();
  } else {
    console.log("width < 768");
  }
</script>

  
</aside>
        
      </div>
      <footer id="footer">
  
  <div class="outer">
    <div id="footer-info" class="inner">
      &copy; 2024 ShengYu<br>
      Powered by <a href="http://hexo.io/" rel="external nofollow noreferrer" target="_blank">Hexo</a>
    </div>
  </div>
</footer>
    </div>
    <nav id="mobile-nav">
  
    <a href="/" class="mobile-nav-link">首頁</a>
  
    <a href="/categories/Python教學/" class="mobile-nav-link">Python教學</a>
  
    <a href="/categories/C-C-教學/" class="mobile-nav-link">C/C++教學</a>
  
    <a href="/archives" class="mobile-nav-link">歸檔</a>
  
    <a href="/tags" class="mobile-nav-link">標籤</a>
  
    <a href="/projects" class="mobile-nav-link">Projects</a>
  
    <a href="/about" class="mobile-nav-link">關於我</a>
  
</nav>
    

<script src="//ajax.googleapis.com/ajax/libs/jquery/2.0.3/jquery.min.js"></script>


  <link rel="stylesheet" href="/fancybox/jquery.fancybox.css">
  <script src="/fancybox/jquery.fancybox.pack.js"></script>


<script src="/js/script.js"></script>



  </div>
</body>
</html>